examples/atmega328p/src/main.rs

```diff
@@ -0,0 +1,93 @@
+#![no_std]
+#![no_main]
+#![feature(abi_avr_interrupt)]
+
+use core::cell::Cell;
+
+use avr_device::interrupt::{self, Mutex};
+
+static LED_STATE: Mutex<Cell<bool>> = Mutex::new(Cell::new(true));
+
+// Our panic handler. It hangs forever while making the PD3 pin blink
+#[cfg(not(doc))]
+#[panic_handler]
+fn panic(_info: &core::panic::PanicInfo) -> ! {
+    // disable interrupts - firmware has panicked so no ISRs should continue running
+    avr_device::interrupt::disable();
+
+    // get the peripherals so we can access serial and the LED.
+    //
+    // SAFETY: Because main() already has references to the peripherals this is an unsafe
+    // operation - but because no other code can run after the panic handler was called,
+    // we know it is okay.
+    let dp = unsafe { avr_device::atmega328p::Peripherals::steal() };
+
+    loop {
+        avr_device::asm::delay_cycles(1_000_000);
+        dp.PORTD.portd.write(|w| w.pd3().set_bit());
+        avr_device::asm::delay_cycles(1_000_000);
+        dp.PORTD.portd.write(|w| w.pd3().clear_bit());
+    }
+}
+
+#[avr_device::interrupt(atmega328p)]
+fn TIMER0_OVF() {
+    // This interrupt should raise every (1024*255)/16MHz s ≈ 0.01s
+    // We then count 61 times to approximate 1s.
+    // XXX: this is a really bad way to count time
+
+    static mut OVF_COUNTER: u16 = 0;
+    const ROLLOVER: u16 = 61;
+
+    *OVF_COUNTER = OVF_COUNTER.wrapping_add(1);
+    if *OVF_COUNTER > ROLLOVER {
+        *OVF_COUNTER = 0;
+
+        interrupt::free(|cs| {
+            LED_STATE.borrow(cs).set(!LED_STATE.borrow(cs).get());
+        });
+    }
+}
+
+#[avr_device::entry]
+fn main() -> ! {
+    let dp = avr_device::atmega328p::Peripherals::take().unwrap();
+
+    // The following should be a NO-op
+    dp.TC0.tccr0a.write(|w| w.wgm0().bits(0));
+
+    // Divide by 1024 -> 16MHz/1024 = 15.6kHz
+    dp.TC0.tccr0b.write(|w| w.cs0().prescale_1024());
+    // Enable overflow interrupts
+    dp.TC0.timsk0.write(|w| w.toie0().set_bit());
+
+    // Make pd2 and pd3 outputs
+    // We use modify in order not to change the other bits
+    dp.PORTD.ddrd.modify(|_, w| w.pd2().set_bit());
+    dp.PORTD.ddrd.modify(|_, w| w.pd3().set_bit());
+
+    unsafe {
+        avr_device::interrupt::enable();
+    }
```

Comment:
`// SAFETY:` comment explaining that we are allowed to enable interrupts here because we know we are not in a critical section.